org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ProcessTaskManager.java

```diff
@@ -46,159 +51,131 @@ public class ProcessTaskManager {
 		try {
-			wait(250);
-		} catch (InterruptedException ignore) {
-			// ignore
+			this.units.put(newElement);
+		} catch (InterruptedException interrupt) {
+			throw new RuntimeException(interrupt);
 		}
-		this.sleepCount = 0;
 	}
 
-	this.units[this.availableIndex++] = newElement;
-	if (this.availableIndex >= this.size)
-		this.availableIndex = 0;
-	if (this.sleepCount <= -1)
-		notify(); // wake up writing thread to accept next unit - could be the last one - must avoid deadlock
-}
+	private Object lastSignal = null;
 
-public CompilationUnitDeclaration removeNextUnit() throws Error {
-	CompilationUnitDeclaration next = null;
-	boolean yield = false;
-	synchronized (this) {
-		next = this.units[this.currentIndex];
-		if (next == null || this.caughtException != null) {
-			do {
-				if (this.processingTask == null) {
-					if (this.caughtException != null) {
-						// rethrow the caught exception from the processingThread in the main compiler thread
-						if (this.caughtException instanceof Error)
-							throw (Error) this.caughtException;
-						throw (RuntimeException) this.caughtException;
-					}
-					return null;
+	/** returns null when no more elements can be expected **/
+	public Collection<CompilationUnitDeclaration> removeNextUnits() throws Error {
+		List<CompilationUnitDeclaration> elements = new ArrayList<>();
+		do {
+			Object next = this.lastSignal;
+			this.lastSignal = null;
+			try {
+				// wait until at least 1 element is available:
+				if (next ==null) {
+					next = this.units.take();
 				}
-				//System.out.print('r');
-				//if (this.sleepCount > 0) throw new IllegalStateException(Integer.valueOf(this.sleepCount).toString());
-				this.sleepCount = -1;
-				try {
-					wait(100);
-				} catch (InterruptedException ignore) {
-					// ignore
+				while (next instanceof CompilationUnitDeclaration cu) {
+					elements.add(cu);
+					// optionally read more elements if already available:
+					next = this.units.poll();
 				}
-				this.sleepCount = 0;
-				next = this.units[this.currentIndex];
-			} while (next == null);
-		}
-
-		this.units[this.currentIndex++] = null;
-		if (this.currentIndex >= this.size)
-			this.currentIndex = 0;
-		if (this.sleepCount >= 1 && ++this.sleepCount > 4) {
-			notify(); // wake up processing thread to add next unit but only after removing some elements first
-			yield = this.sleepCount > 8;
-		}
+				if (!elements.isEmpty()) {
+					if (next !=null) {
+						// defer any stop signal until all CompilationUnitDeclaration read
+						this.lastSignal= next;
+					}
+					return elements;
+				}
+			} catch (InterruptedException interrupt) {
+				throw new RuntimeException(interrupt);
+			}
+			if (next instanceof Error error) {
+				throw error;
+			}
+			if (next instanceof RuntimeException runtimeException) {
+				throw runtimeException;
+			}
+			if (next == STOP_SIGNAL) {
+				return null;
+			}
+			throw new IllegalStateException(String.valueOf(next));
```

Comment:
Could we have a speaking exception message like `"Received unexpected element to process: " + next`

Reply:
we should never see it, but ok.